cloud: fix gcs to use resuming reader

adityamaru opened the pull request:
This change does a few things:
1. gcs_storage was not returning a resuming reader as a result of
which the `Read` method of the resuming reader that contains logic
to retry on certain kinds of errors was not being invoked.

2. Changes the resuming reader to take a storage specific function
that can define what errors are retryable in the resuming reader.
All storage providers use the same deciding function at the moment
and so behavior is unchanged.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65967)
<!-- Reviewable:end -->

dt commented on pkg/storage/cloud/gcp/gcs_storage.go:
```diff
@@ -189,7 +190,7 @@ func (g *gcsStorage) ReadFileAt(
 		}
 		return nil, 0, err
 	}
-	return r.Reader, r.Reader.(*gcs.Reader).Attrs.Size, nil
```
Ooff.

adityamaru committed:
cloud: fix gcs to resuming reader

This change does a few things:

1. gcs_storage was not returning a resuming reader as a result of
which the Read method of the resuming reader that contains logic
to retry on certain kinds of errors was not being invoked.

2, Changes the resuming reader to take a storage specific function
that can define what errors are retryable in the resuming reader.
All storage providers use the same deciding function at the moment
and so behavior is unchanged.

Release note: None

adityamaru committed:
[DNM] cloud: adds GOAWAY error to resumable http errors

This change teaches the resumable reader to check for a
GOAWAY error on Read, and treats it as a resumable error.
The actual error thrown is http.http2GoAwayError, but since
it is not exported in the http pkg, we rely on string comparison.

This commit will not be merged at the moment.

Release note: None

adityamaru commented:
Closing this since we have fixed the resuming reader in https://github.com/cockroachdb/cockroach/pull/66149 and are going to be bumping the google SDK to pull in an SDK level fix for handling GOAWAY errors. Both these changes are also available on 21.1.
